### PR TITLE
refactor(exthost): port definition provider to v2

### PR DESCRIPTION
**Pull request by bryphe**

### CODE REVIEW

File: src/Exthost/Msg.re [hunk 1]
```diff
@@ -240,10 +256,51 @@ module LanguageFeatures = {
...
+    | (
+        "$registerTypeDefinitionSupport",
+        `List([`Int(handle), selectorJson]),
+      ) =>
+      selectorJson
+      |> parseDocumentSelector
+      |> Result.map(selector => {
+           RegisterImplementationSupport({handle, selector})
+         })
+      |> Result.map_error(Json.Decode.string_of_error)
```
Comment (glennsl) [inline src/Exthost/Msg.re:303]: If you're mapping both cases it's usually clearer, and sometimes also shorter, to just use a plain old `switch`

```suggestion
      switch (parseDocumentSelector(selectorJson)) {
      | Ok(selector) => RegisterImplementationSupport({handle, selector})
      | Error(error) => Json.Decode.string_of_error(error)
      }
```

File: src/Exthost/Request.re [hunk 2]
```diff
@@ -110,6 +110,66 @@ module LanguageFeatures = {
...
+
+  module Internal = {
+    let provideDefinitionLink =
+        (~handle, ~resource, ~position, method, client) => {
+      let parser = json => {
+        Json.Decode.(
+          json
+          |> Json.Decode.decode_value(list(DefinitionLink.decode))
+          |> Result.map_error(Json.Decode.string_of_error)
+        );
```
Comment (glennsl) [inline src/Exthost/Request.re:122]: `Json.Decode` is already open here, so no need to qualify `decode_value`:

```suggestion
        Json.Decode.(
          decode_value(list(DefinitionLink.decode), json)
          |> Result.map_error(Json.Decode.string_of_error)
        );
```

File: src/Exthost/Msg.re [hunk 3]
```diff
@@ -240,10 +256,51 @@ module LanguageFeatures = {
       })
     | Unregister({handle: int});
 
+  let parseDocumentSelector = json => {
+    Json.Decode.(
+      json |> Json.Decode.decode_value(list(DocumentFilter.decode))
+    );
```
Comment (glennsl) [inline src/Exthost/Msg.re:262]: `Json.Decode` is already open here, so no need to qualify `decode_value`:

```suggestion
    Json.Decode.(
      json |> decode_value(list(DocumentFilter.decode))
    );
```

**Comment by glennsl** [comment]
Minor note on the title: I think this would be more accurate to classify as a "refactor", as it doesn't add any functionality that doesn't already exist.

**Comment by bryphe** [comment]
Title looks great - thanks @glennsl !

**Pull request merged**